worker/src/ocall_bridge/attestation_ocall_impl.rs

```diff
@@ -0,0 +1,163 @@
+/*
+    Copyright 2019 Supercomputing Systems AG
+    Copyright (C) 2017-2019 Baidu, Inc. All Rights Reserved.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+*/
+
+use crate::ocall_bridge::bridge_api::RemoteAttestationOCall;
+use log::*;
+use sgx_types::*;
+use std::net::{SocketAddr, TcpStream};
+use std::os::unix::io::IntoRawFd;
+use std::ptr;
+
+pub struct RemoteAttestationOCallImpl {
+    // TODO as a member here we need the e-call API trait, so we can use it instead of making the e-call directly
+}
+
+impl RemoteAttestationOCall for RemoteAttestationOCallImpl {
+    fn init_quote(&self) -> (sgx_status_t, sgx_target_info_t, sgx_epid_group_id_t) {
+        // TODO this translation to unsafe C-API should be moved to the EnclaveApi / ECall API
+        let mut ti: sgx_target_info_t = sgx_target_info_t::default();
+        let mut eg: sgx_epid_group_id_t = sgx_epid_group_id_t::default();
+
+        unsafe {
+            let ret_status = sgx_init_quote(
+                &mut ti as *mut sgx_target_info_t,
+                &mut eg as *mut sgx_epid_group_id_t,
+            );
+            (ret_status, ti, eg)
+        }
+    }
+
+    fn get_ias_socket(&self) -> i32 {
+        let port = 443;
+        let hostname = "api.trustedservices.intel.com";
+        let addr = lookup_ipv4(hostname, port);
+        let sock = TcpStream::connect(&addr).expect("[-] Connect tls server failed!");
+
+        sock.into_raw_fd()
+    }
+
+    fn get_quote(
+        &self,
+        revocation_list: Vec<u8>,
+        report: sgx_report_t,
+        quote_type: sgx_quote_sign_type_t,
+        spid: sgx_spid_t,
+        quote_nonce: sgx_quote_nonce_t,
+    ) -> (sgx_status_t, sgx_report_t, Vec<u8>) {
```

Comment:
The intention here was to preserve as much interface logic as possible. I've briefly thought about making it a result, but I hesitated because `sgx_status_t` is not only an error type, but also signals success. So when we 'abuse' it as an error, there can be inconsistencies, where the result is an error, but `sgx_status_t` is 'SUCCESS'. Wrapping it into our own error (as you suggest) might mitigate the problem to some degree

Reply:
So I'll introduce a new error type and have these methods return a result instead, I think that's reasonable 👍